torch/_inductor/fbgemm_lowerings.py

```diff
@@ -0,0 +1,342 @@
+from random import gauss, shuffle
+from typing import Callable, List, Optional, Tuple, Union
+
+import sympy
+
+import torch
+from .ir import Buffer, Pointwise, TensorBox
+from .lowering import fallback_handler, is_integer_type, register_lowering
+from .runtime.runtime_utils import next_power_of_2
+from .virtualized import ops, V
+
+
+# pyre-ignore[2,3]
+def dense_idx_to_jagged_idx(batch_idx, seq_idx, offsets_loader, jagged_len):
+    # jagged_len + 1 is used as the upper bound,
+    # because the last sequence length may be zero
+    begin_idx = ops.indirect_indexing(
+        offsets_loader([batch_idx]),
+        jagged_len + 1,
+    )
+    end_idx = offsets_loader([batch_idx + 1])
+    jagged_idx = begin_idx + seq_idx
+    return jagged_idx, end_idx
+
+
+def get_inverse_offsets(
+    offsets: TensorBox,
+    jagged_len: Union[int, sympy.Expr],
+    realize: bool = True,
+) -> TensorBox:
+    """
+    Returns "inverse_offsets" - the inverse of the offsets array.
+    offsets maps batch index (dense) to jagged index (i.e. offset into jagged tensor).
+    inverse_offsets maps jagged index to batch index.
+
+    e.g. for offsets [0, 3, 4, 9, 10] this will return
+    inverse_offsets = [0, 0, 0, 1, 2, 2, 2, 2, 2, 3]
+
+    For the given offsets, the computed inverse_offsets are cached
+    on the first call and reused in the further calls.
+    """
+
+    if hasattr(offsets, "inverse_offsets"):
+        # inverse_offsets are already computed
+        # for these offsets: can reuse
+        return offsets.inverse_offsets
+
+    # ops.bucketize takes offsets.get_name() which doesn't exist on Pointwise
+    # kernels, i.e. we need to realize it before using. In other words, we need
+    # offsets to be in global memory so that we can binary search over the
+    # entire tensor
+    offsets.realize()
+    device: torch.device = offsets.get_device()
+    dtype: torch.dtype = offsets.get_dtype()
+
+    # pyre-ignore[2,3]
+    def inner_fn(index):
+        idx = index[0]
+        bucket = ops.bucketize(
+            values=ops.index_expr(idx, dtype),
+            offsets_name=offsets.get_name(),
+            offsets_size=offsets.get_size()[0],
+            indexing_dtype=dtype,
+            right=True,
+        )
+        # ops.bucketize above returns 1-based bucket indices,
+        # but we need 0-based, hence we subtract 1 from batch
+        return bucket - 1
+
+    inverse_offsets = Pointwise.create(
+        device=device,
+        dtype=dtype,
+        inner_fn=inner_fn,
+        ranges=[jagged_len],
+    )
+
+    if realize:
+        # "freeze" the node so that it doesn't get inlined downstream.
+        inverse_offsets.realize()
```

Comment:
Why do need to set this ?

Reply:
**TL;DR**: the inverse_offsets code does a binary search that (at least in the past) triton was pretty bad at generating. Additionally, we typically would need to recompute (or re-lookup) the same inverse offset result many times; so instead of recomputing this a lot of times (which triton is pretty bad at), we write it to global memory and then load it when we need it. Adnan Akhundov and I did some benchmarks a while back and found that this was actually a lot faster than trying to fuse it.

Details below:

**inverse_offset computation is duplicated / used multiple times**
the logic for handling dense->jagged conversions is something like this:
* spawn a thread for each element of the output in the jagged format
* **lookup or compute the inverse of the offsets tensor** so we know how to index into dense tensor
* use that index to load your data from the dense tensor and do whatever computation you want

**lookup or compute** - Note that in the typical use case where we're trying to convert to a jagged tensor of shape [J, M], then for each element [j, m] for a given j, the inverse offset computation will be equivalent. So you'll either need to load the same data element multiple times (which is cheap because we're not loading very much data) or duplicate the computation (which is expensive for this particular kernel)

**Second - why the computation is slow**
The inverse offset calculation is done with a binary search.

On one hand, this is expected to be kind of terrible because the entire function is just branching and uncoalesced reads. On the other hand, I'm not really sure why this is worse in triton than in CUDA